google-cloud-storage/src/main/java/com/google/cloud/storage/GapicUnbufferedReadableByteChannel.java

```diff
@@ -153,98 +179,108 @@ public void close() throws IOException {
         leftovers.close();
       }
     } finally {
-      iter.close();
+      readObjectObserver.cancel();
+      drainQueue();
+      ApiExceptions.callAndTranslateApiException(readObjectObserver.cancellation);
+      drainQueue();
+    }
+  }
+
+  private void drainQueue() {
+    IOException ioException = null;
+    while (queue.nonEmpty()) {
+      java.lang.Object queueValue = queue.poll();
+      if (queueValue instanceof ReadObjectResponse) {
+        ReadObjectResponse resp = (ReadObjectResponse) queueValue;
+        try {
+          ResponseContentLifecycleHandle handle = rclm.get(resp);
+          handle.close();
+        } catch (IOException e) {
+          if (ioException == null) {
+            ioException = e;
+          } else if (ioException != e) {
+            ioException.addSuppressed(e);
+          }
+        }
+      } else if (queueValue == EOF_MARKER || queueValue instanceof Throwable) {
+        break;
+      }
     }
   }
 
   ApiFuture<Object> getResult() {
     return result;
   }
 
+  private void ensureStreamOpen() {
+    if (readObjectObserver == null) {
+      readObjectObserver =
+          Retrying.run(
+              retryingDeps,
+              alg,
+              () -> {
+                ReadObjectObserver tmp = new ReadObjectObserver();
+                ReadObjectRequest request = req;
+                long currentFetchOffset = fetchOffset.get();
+                if (request.getReadOffset() != currentFetchOffset) {
+                  request = req.toBuilder().setReadOffset(currentFetchOffset).build();
+                }
+                read.call(request, tmp);
+                ApiExceptions.callAndTranslateApiException(tmp.open);
+                return tmp;
+              },
+              Decoder.identity());
+    }
+  }
+
   private IOException closeWithError(String message) throws IOException {
     close();
     StorageException cause =
         new StorageException(HttpStatusCodes.STATUS_CODE_PRECONDITION_FAILED, message);
     throw new IOException(message, cause);
   }
 
-  private final class LazyServerStreamIterator implements Iterator<ReadObjectResponse>, Closeable {
-    private ServerStream<ReadObjectResponse> serverStream;
-    private Iterator<ReadObjectResponse> responseIterator;
+  private final class ReadObjectObserver extends StateCheckingResponseObserver<ReadObjectResponse> {
```

Comment:
Unfortunately, git thinks this class is a modification of the previous LazyServerStreamIterator except it's a whole new class.

Rather than comparing this class to what was there before, evaluate this class as brand new.